src/alph_puzzle.c

```diff
@@ -0,0 +1,971 @@
+    Unk122_021E6900 unkStruct;
+    MI_CpuFill8(&unkStruct, 0, sizeof(Unk122_021E6900));
+    unkStruct.bgConfig = data->bgConfig;
+    unkStruct.unk8 = 31;
+    unkStruct.unkC = 6;
+    unkStruct.unk4 = 0;
+    unkStruct.unk10 = 25;
+    unkStruct.unk11 = 10;
+    unkStruct.unk12_0 = data->unk4;
+    unkStruct.unk12_4 = 1;
+    unkStruct.unk13 = 0;
+    sub_02016704(data->unk7C, &unkStruct, data->palette);
+}
+
+static AlphPuzzleStates ov110_021E6B94(AlphPuzzleData *data) {
+    AlphPuzzleStates ret;
+    switch (sub_020168F4(data->unk7C)) {
+        case 1:
+            ret = ALPH_PUZZLE_STATE_FADE_OUT;
+            break;
+        case 2:
+            ret = ALPH_PUZZLE_STATE_WAIT_FOR_INPUT;
+            break;
+        default:
+            return ALPH_PUZZLE_STATE_5;
+    }
+    data->unk4 = sub_020169C0(data->unk7C);
+    sub_020169CC(data->unk7C);
+    ClearFrameAndWindow2(&data->window[1], 1);
+    FillWindowPixelBuffer(&data->window[1], 0);
+    ScheduleWindowCopyToVram(&data->window[1]);
+    ScheduleBgTilemapBufferTransfer(data->bgConfig, 0);
+    return ret;
+}
+
+extern s8 ov110_021E6D94[4][2];
+
+void ov110_021E6BEC(AlphPuzzleTile *tile, s16 x, s16 y) {
+    Sprite_SetPositionXY(tile->sprite, ov110_021E6D94[tile->rotation][0] + x, ov110_021E6D94[tile->rotation][1] + y);
+}
+
+void ov110_021E6C18(AlphPuzzleData *data, s16 tileIndex, u8 x, u8 y, u8 rotation) {
+    AlphPuzzleTile *tile = &data->tileGrid[tileIndex];
+    tile->x = x;
+    tile->y = y;
+    tile->rotation = rotation;
+    ov110_021E6BEC(tile, x * 32 + 48, y * 32 + 16);
+    sub_02024818(tile->sprite, (rotation % 4u) * 0x4000);
+}
+
+static void AlphPuzzle_UpdateSelectedTile(AlphPuzzleData *data, u8 tileIndex, BOOL isSelecting) {
+    if (isSelecting) {
+        data->unk1B = tileIndex;
+        data->selectedTile = &data->tileGrid[data->unk1B];
+        Sprite_SetDrawPriority(data->selectedTile->sprite, 0);
+        Sprite_AddPositionXY(data->selectedTile->sprite, -2, -2);
+        Set2dSpriteVisibleFlag(data->unk8C[1], 1);
+        Sprite_SetPositionXY(data->unk8C[1], data->selectedTile->x * 32 + 48, data->selectedTile->y * 32 + 16);
+        ov110_021E6904(data, 2);
+    } else {
+        Sprite_SetDrawPriority(data->selectedTile->sprite, 2);
+        ov110_021E6C18(data, data->unk1B, data->selectedTile->x, data->selectedTile->y, data->selectedTile->rotation);
+        Set2dSpriteVisibleFlag(data->unk8C[1], 0);
+        ov110_021E6904(data, 0);
+        data->unk22 = 0;
+        data->unk23 = 0;
+        data->selectedTile = NULL;
+        data->unk1B = 0;
+    }
+}
+
+typedef struct UnkStruct_021E6D20 {
+    AlphPuzzleData *data;
+    u32 unk4;
+} UnkStruct_021E6D20;
+
+static void ov110_021E6D20(AlphPuzzleData *data) {
+    UnkStruct_021E6D20 *unkStruct = AllocFromHeapAtEnd(data->heapId, sizeof(UnkStruct_021E6D20));
+    MI_CpuFill8(unkStruct, 0, sizeof(AlphPuzzleTile));
+    unkStruct->data = data;
+    CreateSysTask(ov110_021E6D54, unkStruct, 0);
```

Comment:
You can use `UnkStruct_021E6D20` for the task function, you just have to cast the function to a `SysTaskFunc` when calling `CreateSysTask`:
```suggestion
    CreateSysTask((SysTaskFunc)ov110_021E6D54, unkStruct, 0);
```

Reply:
Do we want to do this though? I feel like having it natively as a SysTaskFunc makes more sense